File: 144.binary-tree-preorder-traversal.java
import java.util.ArrayList;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public class Solution {
        public int countUnivalSubtrees(TreeNode root) {
            int[] count = new int[1];
            helper(root, count);
            return count[0];
        }

        public boolean helper(TreeNode root, int[] count) {
            if (root == null)
                return true;
            boolean left = helper(root.left, count);
            boolean right = helper(root.right, count);
            if (left && right) {
                if (root.left != null && root.val != root.left.val)
                    return false;
                if (root.right != null && root.val != root.right.val)
                    return false;
                count[0]++;
                return true;
            }
            return false;
        }
    }
}

File: 144.binary-tree-preorder-traversal.js
/*
 * @lc app=leetcode id=144 lang=javascript
 *
 * [144] Binary Tree Preorder Traversal
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function (root) {
  let stack = [];
  let result = [];
  if (!root) {
    return result;
  }

  stack.unshift(root);

  while (stack.length) {
    let node = stack.shift();
    result.push(node.val);
    if (node.right) {
      stack.unshift(node.right);
    }

    if (node.left) {
      stack.unshift(node.left);
    }
  }

  return result;
};
